test_application/src/main.rs

```diff
@@ -0,0 +1,272 @@
+    let mut next_action: NextActionChannels = Vec::new();
+
+    let mut participants: Vec<ParticipantIdentifier> = Vec::new();
+    for _n in 0..num_workers {
+        participants.push(ParticipantIdentifier::random(rng));
+    }
+
+    info!("Spawning {num_workers} worker threads");
+    // Spawn participants. Link them with channels.
+    for i in 0..num_workers {
+        let id = participants[i];
+        let mut other_ids = participants.clone();
+        other_ids.remove(i);
+
+        let (worker_message_tx, worker_message_rx) = unbounded::<Message>();
+        let (next_action_tx, next_action_rx) = unbounded::<(SubProtocol, Identifier)>();
+        next_action.push(next_action_tx);
+        worker_messages.insert(id, worker_message_tx);
+
+        let config = ParticipantConfig { id, other_ids };
+        let outgoing = outgoing_tx.clone();
+
+        thread::spawn(|| participant_worker(config, next_action_rx, worker_message_rx, outgoing));
+    }
+
+    do_tss_ecdsa(worker_messages, next_action, workers_rx, num_workers)?;
+
+    Ok(())
+}
+
+fn do_tss_ecdsa(
+    workers_message: WorkerChannelsMessage,
+    next_action: NextActionChannels,
+    workers_rx: Receiver<MessageFromWorker>,
+    num_workers: usize,
+) -> anyhow::Result<()> {
+    info!("Starting key generation.");
+    start_subprotocol(&next_action, SubProtocol::StartKeyGen)?;
+    route_worker_messages(&workers_message, &workers_rx, num_workers)?;
+    info!("Finished key generation.");
+
+    info!("Starting aux-info.");
+    start_subprotocol(&next_action, SubProtocol::StartAuxInfo)?;
+    route_worker_messages(&workers_message, &workers_rx, num_workers)?;
+    info!("Finished aux-info.");
+
+    info!("Starting presign.");
+    start_subprotocol(&next_action, SubProtocol::StartPresign)?;
+    route_worker_messages(&workers_message, &workers_rx, num_workers)?;
+    info!("Finished presign.");
+
+    info!("Starting sign.");
+    start_subprotocol(&next_action, SubProtocol::Sign)?;
+    route_worker_messages(&workers_message, &workers_rx, num_workers)?;
+    info!("Finished sign.");
+
+    Ok(())
+}
+
+fn start_subprotocol(
+    next_action: &NextActionChannels,
+    sub_protocol: SubProtocol,
+) -> anyhow::Result<()> {
+    debug!("Starting sub-protocol: {:?}", sub_protocol);
+    let rng = &mut thread_rng();
+    let sid = Identifier::random(rng);
+
+    for worker in next_action {
+        worker.send((sub_protocol, sid))?;
+    }
+
+    Ok(())
+}
+
+fn route_worker_messages(
+    send_to_worker: &WorkerChannelsMessage,
+    from_workers: &Receiver<MessageFromWorker>,
+    num_workers: usize,
+) -> anyhow::Result<()> {
+    // # of workers who finished the current sub_protocol.
+    let mut sub_protocol_ended = 0;
+
+    for message in from_workers {
+        debug!("Received messages from a worker.");
+        match message {
+            MessageFromWorker::FinishedRound(messages) => {
+                if messages.is_empty() {
```

Comment:
Workers appear to return empty vectors of messages sometimes. We should wrap the `Vec<Message>` in an option instead of relying on the empty behavior, this way users don't have to check if the vec is empty. This is specially important when we will be sending these messages over the network.